### PR TITLE
Health Check

### PR DESCRIPTION
**Pull request by ZairMahmood**
Adding a method to quickly get health of sand for building health endpoints

### CODE REVIEW

**Comment by raghunandan-somaraju** [comment]
@ZairMahmood why do we need to check the health of Sand Service? There will be ops checks on it anyways.

**Comment by ZairMahmood** [comment]
@raghunandan-somaraju In the even that this service is being brought back up by ops, each service has the option to implement this so it knows that sand is down when they hit their own health endpoint

File: sand_python/sand_service.py [hunk 1]
```diff
@@ -88,7 +88,16 @@ def validate_request(self, request_headers, opts={}):
         validation_resp = self.__validate_with_sand(client_token, service_token, scopes, opts)
         self.cache.set(self.__get_client_token_cache_key(client_token, scopes), validation_resp, self.__get_cache_expiry_secs(validation_resp))
         return validation_resp
-
+    
+    # Get health status of sand
+    # Useful for building health endpoints
+    def health(self):
```
Comment (raghunandan-somaraju) [inline sand_python/sand_service.py]: We need to call it something more explicit. Maybe get_sand_service_health? Every service will have a "health" method of its own.

Comment (ZairMahmood) [reply sand_python/sand_service.py]: updated

**Comment by johnwu96822** [comment]
A service's health check is a HTTP Get request on the service's health address. In Go we have a foundation-go library to do it (see https://github.com/coupa/foundation-go/blob/master/health/web_check.go), which checks health on any given address.

What I am trying to say is that, doing this health check does not seem like a job for Sand client/service, since the health check function itself is too general (a GET request on a given address?). Instead if you have some library that supports generic microservice standard functions (like foundation-go), it belongs there. Also the library should not assume what the health path is, which may or may not change when we change Sand to fit the microservice standard in the future.

**Comment by johnwu96822** [comment]
And to Raghu's point that services don't need to check Sand health, he is correct. Clients or services just call their usual methods of getting token and verifying token, and they will handle if Sand is down. Nobody does a health check first and then call the get or verify token. Health check access is only useful for implementing the detailed health for our microservice standard. But again this should not be the responsible of the client library.

**Comment by ZairMahmood** [comment]
I primarily placed this here because the client library knows what URL it's accessing. To your point, the client is still calling their usual methods. The service itself never uses it and a microservice should never check the health of the service before getting or verifying the token. Another plus in my opinion is that if the health endpoint does change, it can be changed here once rather than every microservice that makes use of sand-python but I do agree this may not be the best place to put the sand health check and I am open to closing this.

**Comment by johnwu96822** [comment]
@ZairMahmood It's the same reason that both the token path and token verification path are configurable in the library, it's because the library can't assume what they are. Same applies to the health endpoint.

To your second point, if the health endpoint is kept in this library and is changed, all microservices still need to upgrade their version of sand-python. Which is still a code change and re-deploy. I guess nothing comes free in terms of changes :)

I'd say we can close this, if you and @raghunandan-somaraju agree.

**Comment by raghunandan-somaraju** [comment]
@ZairMahmood lets close this for now. When we get a better usage example, we can do it at that time.

**Pull request closed without merging**